File: backend/app/schemas/risk.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

from app.schemas.common import SourceMeta


class RiskCheckRequest(BaseModel):
    hypothesis_id: str
    backtest_id: str | None = None
    account_equity: float = 10000
    risk_per_trade: float = 0.011
    position_size: float = 0.65
    entry_price: float = 3446
    stop_loss: float = 3392
    take_profit: float | None = 3544


class RiskItem(BaseModel):
    name: str
    status: str
    message: str


class RiskCheckRead(SourceMeta):
    id: str
    workflow_id: str
    hypothesis_id: str
    backtest_id: str | None = None
    decision: str
    checks: list[RiskItem]
    warnings: list[str]
    block_reasons: list[str]
    created_at: datetime


class RiskCheckResponse(BaseModel):
    ok: Literal[True] = True
    data: RiskCheckRead


class RiskCheckListResponse(BaseModel):
    ok: Literal[True] = True
    data: list[RiskCheckRead]

File: src/api/risk.ts
import { apiJson, apiRequest } from "./client";

export type RiskCheckRequest = {
  hypothesis_id: string;
  backtest_id?: string | null;
  account_equity?: number | null;
  risk_per_trade?: number | null;
  position_size?: number | null;
  entry_price?: number | null;
  stop_loss?: number | null;
  take_profit?: number | null;
};

export type RiskItem = {
  name: string;
  status: string;
  message: string;
  threshold?: string | null;
  actual?: string | null;
};

export type RiskCheck = {
  id: string;
  workflow_id: string;
  hypothesis_id: string;
  backtest_id?: string | null;
  decision: "APPROVED" | "CONDITIONAL" | "REJECTED" | "PASS" | "WARNING" | "BLOCK" | string;
  status: "APPROVED" | "CONDITIONAL" | "REJECTED" | string;
  risk_level: string;
  risk_score: number;
  account_equity: number;
  risk_per_trade: number;
  position_size: number;
  entry_price: number;
  stop_loss: number;
  take_profit?: number | null;
  max_loss: number;
  reward_risk: number;
  leverage: number;
  execution_mode: string;
  live_trading_enabled: boolean;
  checks: RiskItem[];
  rule_results: RiskItem[];
  blocks: string[];
  warnings: string[];
  block_reasons: string[];
  market_data_status: "live" | "unavailable" | string;
  created_at: string;
  is_mock: boolean;
  source: string;
};

export function runRiskCheck(payload: RiskCheckRequest) {
  return apiJson<RiskCheck>("/api/risk/checks", payload);
}

export function fetchRiskCheck(id: string, signal?: AbortSignal) {
  return apiRequest<RiskCheck>(`/api/risk/checks/${encodeURIComponent(id)}`, { signal });
}

export function fetchRiskRules(signal?: AbortSignal) {
  return apiRequest<Array<{ name: string; label: string; threshold: string; source: string }>>("/api/risk/rules", { signal });
}
